Tools/gsc/gsc.py

```diff
@@ -0,0 +1,219 @@
+    if entrypoint_len > 1:
+        last_bin_arg = entrypoint_len
+        binary_arguments = '"' + '", "'.join(entrypoint[1: last_bin_arg]) + '"'
+    else:
+        last_bin_arg = 0
+        binary_arguments = ''
+
+    # Place the remaining optional arguments previously specified as command, in the
+    # new command. This is necessary, since the first element of the command may be the
+    # binary of the resulting image.
+    remaining_args = '"' + '", "'.join(entrypoint[last_bin_arg + 1 : ]) + '"'
+    cmd = remaining_args if len(entrypoint) > last_bin_arg + 1 else ''
+
+    return binary, binary_arguments, cmd
+
+def prepare_env(base_image, image, args, user_manifests):
+    env = jinja2.Environment(loader=jinja2.FileSystemLoader('templates/'))
+
+    env.globals.update({'args': args})
+    config = load_config('config.yaml')
+    env.globals.update(config)
+
+    # Image names follow the format distro/package:tag
+    image_re = re.match(r'([^:]*)(:?)(.*)', image[image.rfind('/')+1:])
+
+    # Extract binary and command from base_image
+    binary, binary_arguments, cmd = extract_binary_cmd_from_image_config(
+                                                            base_image.attrs['Config'])
+
+    working_dir = base_image.attrs['Config']['WorkingDir']
+
+    env.globals.update({
+            'app_image' : image,
+            'app' : image_re.group(1),
+            'binary_arguments' : binary_arguments,
+            'cmd' : cmd,
+            'working_dir': working_dir,
+            'user_manifests': ' '.join([os.path.basename(manifest)
+                                        for manifest in user_manifests[1:]])
+            })
+
+    return env, binary
+
+# Build graphenized docker image. args has to follow [<options>] <base_image> <app.manifest>
+# [<app2.manifest> ...].
+def gsc_build(args):
+
+    image = args.image
+    user_manifests = args.manifests
+
+    docker_socket = docker.DockerClient(base_url='unix://var/run/docker.sock')
+
+    try:
+        docker_socket.images.get(gsc_image_name(image))
+        print(f'Image {gsc_image_name(image)} already exists, no gsc build required.')
+        sys.exit(0)
+
+    except (docker.errors.ImageNotFound, docker.errors.APIError):
+        try:
+            base_image = docker_socket.images.get(image)
+        except (docker.errors.ImageNotFound, docker.errors.APIError):
+            print(f'Unable to find base image {image}')
+            sys.exit(1)
+
+        print(f'Building graphenized image from base image {image}')
+
+        env, binary = prepare_env(base_image, image, args,
+                                                        user_manifests)
+
+        prepare_build_context(image, user_manifests, env, binary)
+
+        docker_api = docker.APIClient(base_url='unix://var/run/docker.sock')
+        # docker build returns stream of json output
+        stream = docker_api.build(path='gsc-' + image, tag=gsc_image_name(image))
```

Comment:
It might be useful to add a gsc build option to rebuild entire gsc docker image from scratch (without using cached docker images). For example, this will be useful in cases where graphene runtime has to be rebuild and gsc docker image needs to use updated graphene runtimes.
setting `nocache` to `True` in docker API `build()` will rebuild docker image without using cache. This could be set based on gsc build option.
`docker_api.build(path='gsc-' + image, tag=gsc_image_name(image), 
 nocache=True)`